src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,169 @@
+use std::ops::{Add, AddAssign, Range, Sub};
+use std::fmt::Debug;
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct RangeAllocator<T> {
+    /// The range this allocator covers.
+    range: Range<T>,
+    /// A Vec of ranges in this heap which are unused.
+    /// Must be ordered with ascending range start to permit short circuiting allocation.
+    /// No two ranges in this vec may overlap.
+    empty_ranges: Vec<Range<T>>,
+}
+
+impl<T> RangeAllocator<T>
+    where T: Clone + Copy + Add<Output=T> + AddAssign + Sub<Output=T> + Eq + PartialOrd + Debug
+{
+    pub fn new(range: Range<T>) -> Self {
+        RangeAllocator {
+            range: range.clone(),
+            empty_ranges: vec![range.clone()],
+        }
+    }
+
+    pub fn range(&self) -> Range<T> {
+        self.range.clone()
+    }
+
+    pub fn allocate_range(&mut self, length: T) -> Option<Range<T>> {
+        let mut best_range_tuple: Option<(usize, Range<T>)> = None;
+        for range in self.empty_ranges.iter().cloned().enumerate() {
+            if range.1.end - range.1.start < length {
+                continue;
+            }
+            match best_range_tuple {
+                Some((ref mut index, ref mut best_range)) => {
+                    // We want to find the best fit for this allocation in order to reduce 
+                    // memory fragmentation, so we'll look for it here.
+                    if range.1.end - range.1.start < best_range.end - best_range.start {
+                        *index = range.0;
+                        *best_range = range.1.clone();
+                        // Found a perfect fit, so stop looking.
+                        if range.1.end - range.1.start == length {
+                            break;
+                        }
+                    }
+                },
+                None => { 
+                    best_range_tuple = Some(range.clone());
+                    // Found a perfect fit, so stop looking.
+                    if range.1.end - range.1.start == length {
+                        break;
+                    }
+                }
+            }
+        }
+        best_range_tuple.map(|range| {
+            if range.1.end - range.1.start == length {
+                self.empty_ranges.remove(range.0);
+            } else {
+                self.empty_ranges[range.0].start += length;
+            }
+            range.1.start..(range.1.start + length)
+        })
+    }
+
+    pub fn free_range(&mut self, range: Range<T>) -> bool {
```

Comment:
Why return a bool here?

Reply:
why return anything at all?

Reply:
Because the free could fail and it's important to communicate that.

Reply:
The current use cases seem to all ignore the return value.